packages/pkl.experimental.syntax/TypeNode.pkl

```diff
@@ -65,8 +66,17 @@ class ConstrainedTypeNode extends TypeNode {
       renderedUnderlyingType + renderConstraints(currentIndent)
 }
 
+class UnionDefaultType extends Node {
+  typeNode: TypeNode
+
+  function render(currentIndent: String) = "*\(typeNode.render(currentIndent))"
+}
+
 class UnionTypeNode extends TypeNode {
-  members: Listing<TypeNode>
+  members: Listing<TypeNode|UnionDefaultType>(atMostOneUnionDefaultType)
+
+  const local atMostOneUnionDefaultType = (members: Listing<TypeNode|UnionDefaultType>) ->
```

Comment:
```suggestion
  members: Listing<TypeNode|UnionDefaultType>(hasAtMostOneUnionDefaultType)

  const local hasAtMostOneUnionDefaultType = (members: Listing<TypeNode|UnionDefaultType>) ->
```

Nit: constraints should read like a sentence, with an implied "this"; e.g. "this has at most one union default type"